File: secondo_anno/algoritmi/gare/1/snow.rs
use std::collections::HashMap;
include!("std.rs");

fn max_rep(nums: &Vec<usize>) -> usize {
    if nums.len()==0 { return 0; }
    let mut index = HashMap::new();
    let mut l: usize = 0;
    let mut count: usize = 0;
    for r in 0..nums.len() {
        if !index.contains_key(&nums[r]) || index.get(&nums[r]).unwrap() < &l {
            index.insert(&nums[r], r+1);
            count = if count < r-l { r-l } else { count };
        }
        else {
            l = if index[&nums[r]] > l { index[&nums[r]] } else { l };
            index.insert(&nums[r], r+1);
        }
    }
    count+1
}


fn main() {
    let samples= get_type::<isize>().unwrap();
    for _j in 0..samples {
        let size = get_type::<usize>().unwrap();
        let mut nums = vec![0 as usize; size];
        for i in 0..size {
            nums[i] = get_type().unwrap();
        }
        println!("{}", max_rep(&nums));
    }

}    

File: algoritmi/gare/1/snow.cpp
#include <iostream>
#include <unordered_map>

int longest(int *nums, int n) {
    if (n==0) return 0;
    std::unordered_map<int, unsigned int> index;
    unsigned int l = 0, count = 0;
    for (unsigned int r=0; r<n; r++) {
        if (!index[nums[r]] || index[nums[r]] < l) {
            index[nums[r]] = r+1;
            count = count < r - l ? r - l : count;
        }
        else {
            l = index[nums[r]] > l ? index[nums[r]] : l;
            index[nums[r]] = r+1;
        }
    }
    return count +1;
}

int main() {
    int samples;
    std::cin >> samples;
    for (int i=0; i<samples; i++) {
        int n;
        std::cin >> n;
        int* nums= new int[n];
        for (int j=0; j<n; j++) std::cin >> nums[j];
        std::cout << longest(nums, n) << std::endl;
        delete[] nums;
    }
    return 0;
}

